feat: Prints firmware revision in Settings

waltonzt opened the pull request:
Makes the firmware version in settings clickable and link it to the particular firmware's commit.

Related to: https://github.com/trezor/trezor-suite/issues/2917

szymonlesisz commented on packages/suite/src/views/settings/device/index.tsx:
```diff
@@ -199,13 +217,32 @@ const Settings = () => {
-                                    values={{ version: getFwVersion(device) }}
+                                    values={{
+                                        version: (
+                                            <VersionTooltip content={getFwRevision(device) || ''}>
```
assign `const revision = getFwRevision(device);` and use it here and in line 228, fallback is not necessary (?)

waltonzt replied:
14d766f

keraf commented on packages/suite/src/views/settings/device/index.tsx:
```diff
@@ -199,13 +217,32 @@ const Settings = () => {
+                                                        icon="EXTERNAL_LINK"
+                                                        alignIcon="right"
+                                                    >
+                                                        {getFwVersion(device)}
```
This function is called 3 times here, would be best to assign the result to a variable an use it here.

waltonzt replied:
`getFwVersion` is only used once in this file. `getFwRevision` is used twice so I am moving that to a variable.

keraf replied:
Indeed, sorry! I should start wearing my glasses more often :smile:

waltonzt replied:
14d766f

szymonlesisz commented on packages/suite/src/views/settings/device/index.tsx:
```diff
@@ -199,13 +217,32 @@ const Settings = () => {
+                                    values={{
+                                        version: (
+                                            <VersionTooltip content={getFwRevision(device) || ''}>
+                                                <VersionLink
```
we do have this special component for external links, it handles the cases when TOR is enabled.
you can get inspiration from here (it also have this fancy github logo :])
https://github.com/trezor/trezor-suite/blob/develop/packages/suite/src/components/firmware/Initial.tsx#L148-L152

waltonzt replied:
That special component is not used for displaying suite version. Will likely want to change that at some point I would think.

https://github.com/trezor/trezor-suite/blob/9e8c13a5b2b666a2a06c7703fd3aa8e43dde67f9/packages/suite/src/views/settings/index.tsx#L362

keraf replied:
As mentioned in the other comment, `TrezorLink` was designed for trezor.io links.

szymonlesisz commented on packages/suite/src/views/settings/device/index.tsx:
```diff
@@ -199,13 +217,32 @@ const Settings = () => {
+                                            <VersionTooltip content={getFwRevision(device) || ''}>
+                                                <VersionLink
+                                                    target="_blank"
+                                                    href={`https://github.com/trezor/trezor-firmware/commit/${getFwRevision(
```
move this link to https://github.com/trezor/trezor-suite/blob/develop/packages/suite/src/constants/suite/urls.ts

waltonzt replied:
The suite version url is not a constant as well. Would you want me to move it there as well?

https://github.com/trezor/trezor-suite/blob/9e8c13a5b2b666a2a06c7703fd3aa8e43dde67f9/packages/suite/src/views/settings/index.tsx#L364

szymonlesisz replied:
well... i guess it should be. 
@keraf is there any reason why `TrezorLink` component is not used there (in settings index)?
my guess is that TrezorLink component wasn't there yet

keraf replied:
The `TrezorLink` component was designed for `trezor.io` links and to be converted as Tor links when that feature is enabled. Not sure if there's a specific component for other links but it looks all good here for me.

waltonzt replied:
Okay so it sounds like I should move the url to the constants but keep the link the same?

szymonlesisz replied:
ok, so we have 4 different approaches here and i think it should be unified at some point
`TrezorLink` - for handling tor links
`ExternalLink` - which is using TrezorLink
`@trezor/components/Link` - used for external links, not trezor.io. it's covering cases like "noreferrer" or stopPropagation
`styled.a` - used for external links, not trezor.io, not covering anything

waltonzt replied:
14d766f

szymonlesisz replied:
great, thanks
i've addressed this link unification issue https://github.com/trezor/trezor-suite/issues/3633

keraf commented on packages/suite/src/utils/suite/device.ts:
```diff
@@ -121,6 +121,12 @@ export const getFwVersion = (device: AcquiredDevice) => {
 
+export const getFwRevision = (device: AcquiredDevice) => {
+    const { features } = device;
+    const rev = Buffer.from(features.revision, 'hex');
```
We are currently looking to reduce the amount of polyfills required in Suite and `Buffer` is one of them. There might maybe be some utils in Suite to convert from hex to string but I can't remember where (if it exists) from the top of my head. If it doesn't exist, it should be fairly trivial to implement.

szymonlesisz replied:
tbh. i don't think we will ever get rid of `Buffer` polyfil, it's used in so many places and dependecies...

keraf replied:
I'd love to reduce the amount of polyfills we have in the project, especially given the alternatives. Not allowing new uses of `Buffer` in incoming changes is a good step towards that direction.

szymonlesisz replied:
whats your plan to reduce usage of `Buffer` in suite dependencies? 
like crypto/coins libs (btc, eth, xrp...), address validators libs, etc?
basically what i'm trying to say is that `Buffer` is so essential to work with crypto/hashes (metadata, crypto addresses, parsing tx hashes..) that in my opinion it is here to stay forever in this project :)

waltonzt replied:
Hows something like this:

```
export const getFwRevision = (device: AcquiredDevice) => hex2String(device.features.revision)

export const hex2String = (hex: string) => {
    let str = '';
    for (let i = 0; i < hex.length; i += 2) {
        str += String.fromCharCode(parseInt(hex.substr(i, 2), 16));
    }
    return str;
};
```

waltonzt replied:
14d766f

keraf replied:
@szymonlesisz I still have to evaluate how widespread the usage of Node APIs (which are polyfilled for the front-end) is. I'm aware Buffer is useful in our project but there are ways to do most of it (if not all) using APIs available in the browser. Not sure how long that's gonna take but I wouldn't want to introduce new code that relies on this polyfill in the project.

@waltonzt Looks good, but I would suggest to move the `hex2String` function in a different utils file such as `suite/string.ts` as it is quite generic.

szymonlesisz replied:
i understand the intention and i have nothing against reducing polyfils, yet i don't understand how you want deal with that regarding dependencies while all essential crypto modules are using `Buffer`

keraf reviewed (changes requested):
Thanks for the contribution! I've added two comments for some changes.

szymonlesisz commented on packages/suite/src/utils/suite/device.ts:
```diff
@@ -121,6 +121,12 @@ export const getFwVersion = (device: AcquiredDevice) => {
     return `${features.major_version}.${features.minor_version}.${features.patch_version}`;
 };
 
+export const getFwRevision = (device: AcquiredDevice) => {
```
```suggestion
export const getFwRevision = (device: AcquiredDevice) => Buffer.from(device.features.revision, 'hex').toString();
```
i would prefer this one liner.
looks cleaner imo and doesnt assign any variables

szymonlesisz replied:
on second thought, you don't need to create utility for that since it's used only in 1 place.
it could be used/assigned directly in settings/device view

waltonzt replied:
14d766f

szymonlesisz replied:
i'm still not convinced about not using the Buffer, but i leave it for @keraf to decide, he is the pipeline master :)

szymonlesisz reviewed (changes requested):
Thanks for contribution!
i got few comments regarding links

keraf commented:
Hi @waltonzt!

Sorry for letting this PR stale, it slipped from my mind and I forgot to get back to it. I discussed the `Buffer` use with @szymonlesisz and we agreed to keep using it for now. Could you please just revert this portion of the code to your initial code and squash the fixups? I'll merge this PR after that :slightly_smiling_face: 

Cheers!

waltonzt commented:
Hey @keraf,

I have reverted that change to use `Buffer` again and squashed my commits into one. Please let me know if you need anything else from me. 

Thanks!

keraf commented:
Closing in favour of #3743, thanks for the contribution!
